tests/h/views/api_groups_test.py

```diff
@@ -10,6 +10,44 @@
 from h.views import api_groups as views
 
 
+@pytest.mark.usefixtures('profile_group_service')
+class TestGroups(object):
+    def test_groups_proxies_to_service(self, pyramid_request, profile_group_service):
+        views.groups(pyramid_request)
+
+        assert profile_group_service.called_once()
```

Comment:
This is a tricky one which I've fallen into myself before, and I wouldn't be surprised if there are instances of this mistake hiding in the code on master too: I think these tests will always pass. The mock library **doesn't have a `called_once()` method** (or `called_once_with()` either). The methods are called `assert_called_once()` and `assert_called_once_with()`. Instead of `assert profile_group_service.called_once()` it should be `profile_group_service.assert_called_once()`.

So why does this test pass rather than crash and fail?

```python
(Pdb) profile_group_service.called_once_with(pyramid_request, document_uri='foo.example.com')
<Mock name='mock.called_once_with()' id='139667476371408'>

(Pdb) pp profile_group_service.baked_potato(cheese=True)
<Mock name='mock.baked_potato()' id='139667476371728'>
```

By default you can call __any__ method on a `Mock` or `MagicMock` object, including ones that "don't exist" such as `called_once()`, or a typo like `asert_called_once()`, etc., and it will return another mock object. And mock objects, like most objects in Python, are truthy. So `assert <some mock object>` passes. Ouch!

Some more explanation of this issue in the autospeccing docs: <http://www.voidspace.org.uk/python/mock/helpers.html#autospeccing>

This is because the mock `profile_group_service` is an unconstrained mock:

```python
svc = mock.Mock()
```

So I think the best remedy here is to create this `svc` object using `mock.create_autospec()` which will constrain the mock object to having only those methods that `ProfileGroupService` has plus mock library methods like `assert_called_once()`.. Or if `create_autospec()` won't work in this case (it has its limitations) then fall back on the `spec` or `spec_set` args to `mock.Mock()`. (In this instance I think `create_autospec()` should work though.)

Despite the fact that our tests do it all over the place, it should probably be a general rule to never create an unconstrained mock.

Reply:
Oh, and I see this pr just _moved_ this existing code from one file to another. So I guess a previous pr introduced this problem and I didn't spot it then.